ruy/mul_params.h

```diff
@@ -200,16 +246,24 @@ template <typename DstScalar>
 struct MulParamsStorage<std::int32_t, DstScalar> final {
   using AccumScalar = std::int32_t;
   static_assert(std::is_integral<DstScalar>::value, "");
-  static_assert(sizeof(DstScalar) < sizeof(AccumScalar), "");
+  static_assert(sizeof(DstScalar) <= sizeof(AccumScalar) / 2, "");
 
   const AccumScalar* bias = nullptr;
   union {
-    const AccumScalar* multiplier_fixedpoint_perchannel = nullptr;
-    AccumScalar multiplier_fixedpoint;
+    const AccumScalar* multiplier_fixedpoint_perchannel;
+    // Let the default multiplier be effecively a multiplication by 1, so that
+    // the matmul behaves as a (saturating) plain integer matmul. Unfortunately
+    // 1 is not exactly representable in fixedpoint with 0 integer bits, but
+    // using the highest representable value is a sufficiently good
+    // approximation: since this specialization of MulParams is for the case
+    // where DstScalar is at least 2x narrower than MulScalar, the values
+    // for which there would be a difference will get saturated anyway.
+    AccumScalar multiplier_fixedpoint = std::numeric_limits<AccumScalar>::max();
   };
   union {
-    const int* multiplier_exponent_perchannel = nullptr;
-    int multiplier_exponent;
+    const int* multiplier_exponent_perchannel;
+    // See the above comment about the default value of multiplier_fixedpoint.
+    int multiplier_exponent = 0;
```

Comment:
I think the existence of this exponent is one of the things that confused me. I saw that there was a scaling factor and so I assumed that this was the part that indicated the scaling factor. Why is this necessary at all? It seems like can express the full range of scalings with just the fixed point, from no scaling (as you've got to the default) down to INT32_MAX -> 1, with a fixedpoint of 1 (which is really 2^-31)

Reply:
Great question!

For multipliers < 1, as you say, technically the multiplier_fixedpoint allows by itself to encode multipliers as small as `2^-31`. However, the relative accuracy of such small multipliers would be low. For example, while we could represent the multiplier `2^-24` as the integer value `128` , the next representable multipler value, represented by the integer `129`, would be `(129/128) * 2^-24`. So if the actual multiplier value that we need to represent is close to `128.5` then we have to round it , introducing a rounding error of relative magnitude close to `0.5 / 128` or `0.4%`. Even if the destination type is `int8`, itself with only 256 representable values so inherently forcing a rounding error of the same order of magnitude, introducing this new rounding error on top of that can be a big deal, particularly as this would manifest itself as a source of `bias`, i.e. an error term whose distribution isn't centered around 0. Indeed, in the uniform (non-per-channel) quantization case, there is this single multiplier value that all of the current layer's output activations get scaled by. Any error on this multiplier itself gets propagated to all of the layer's output values - so the entire distribution of output activations gets shifted. This kind of bias is known to be destructive to end-to-end NN accuracy, as it accumulates across layers.

It is true that in the majority of cases, multipliers will be larger than `2^-24`, closer to `2^-16`, making that issue less important, and really I don't have experimental data showing that this issue matters in practice. But on the other hand, supporting these exponents has negligible overhead, so the incentive to try to remove them is rather low.

For multipliers > 1 , which is a weird thing that we have learned from practical experience we have to support, the need for the exponent is more obvious. While we could remove it by changing the fixed-point format of the multiplier_fixedpoint, that would be an awkward compromise between accuracy and estimating what's the largest multiplier that we need to support. The original estimate based on common sense was that multipliers should never be > 1, but that turned out to fail to account for the need to evaluate incompletely-trained neural networks, among a few other edge cases.

Actually, here is my regret on this topic:
We should probably just have had a single floating-point value for this multiplier. Plain float32. The motivation not to do that was that we wanted this quantized inference stuff to be 100% integer-arithmetic-only, so we could support running on integer-arithmetic-hardware. But now I think we missed the fact that we could still have had the multiplier specified in fixed-point in the API, with the understanding that an implementation could read the mantissa and exponent from that float and use them in pure integer arithmetic, like the current multiplier fields.

Reply:
Interesting, thanks for the background :-) Yeah it just feels weird that when setting these I have tons of options for representing the exact same multiplier.

Reply:
If that helps, consider that you should only use multiplier_fixedpoint values in the range from 2^30 to 2^31-1.
Consider values below 2^30 as "not normalized" (in the same sense as in floating-point denormals).